core/trie/proof.go

```diff
@@ -133,6 +133,206 @@ func transformNode(tri *Trie, parentKey *Key, sNode StorageNode) (*Edge, *Binary
 	return edge, binary, nil
 }
 
+// Checks if there happens at most one split in the merged path
+func pathSplitOccurredCheck(mergedPath []ProofNode, nodeHashes map[felt.Felt]ProofNode) error {
+	splitHappened := false
+	for _, node := range mergedPath {
+		if node.Edge != nil {
+			continue
+		}
+
+		leftHash := node.Binary.LeftHash
+		rightHash := node.Binary.RightHash
+
+		_, leftExists := nodeHashes[*leftHash]
+		_, rightExists := nodeHashes[*rightHash]
+
+		if leftExists && rightExists {
+			if splitHappened {
+				return errors.New("split happened more than once")
+			}
+
+			splitHappened = true
+		}
+	}
+	return nil
+}
+
+func noCycleCheck(node ProofNode, nodeHashes map[felt.Felt]ProofNode, visitedHashes map[felt.Felt]bool, hash hashFunc) error {
+	nodeHash := node.Hash(hash)
+	_, visited := visitedHashes[*nodeHash]
+
+	if visited {
+		return errors.New("circular path in the merged proof")
+	}
+
+	visitedHashes[*nodeHash] = true
+
+	if node.Binary != nil {
+		leftHash := node.Binary.LeftHash
+		rightHash := node.Binary.RightHash
+
+		leftNode, leftExist := nodeHashes[*leftHash]
+		rightNode, rightExist := nodeHashes[*rightHash]
+
+		if leftExist {
+			if err := noCycleCheck(leftNode, nodeHashes, visitedHashes, hash); err != nil {
+				return err
+			}
+		}
+
+		if rightExist {
+			if err := noCycleCheck(rightNode, nodeHashes, visitedHashes, hash); err != nil {
+				return err
+			}
+		}
+	} else if node.Edge != nil {
+		edgeNode, exist := nodeHashes[*node.Edge.Child]
+		if exist {
+			if err := noCycleCheck(edgeNode, nodeHashes, visitedHashes, hash); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func rootNodeExistsCheck(rootHash *felt.Felt, nodeHashes map[felt.Felt]ProofNode) (ProofNode, error) {
+	currNode, rootExists := nodeHashes[*rootHash]
+	if !rootExists {
+		return currNode, errors.New("root hash not found in the merged path")
+	}
+
+	return currNode, nil
+}
+
+// Begins from the root node and traverses the merged proof path
+// Until it finds a split node, adds nodes to commonPath
+// Then continues with traversing left and right paths separetaly
+// Assumes there is no circular paths and the split happens at most once
+func traverseNodes(currNode ProofNode, path *[]ProofNode, nodeHashes map[felt.Felt]ProofNode) {
+	*path = append(*path, currNode)
+
+	if currNode.Binary != nil {
+		expectedLeftHash := currNode.Binary.LeftHash
+		expectedRightHash := currNode.Binary.RightHash
+
+		nodeLeft, leftExist := nodeHashes[*expectedLeftHash]
+		nodeRight, rightExist := nodeHashes[*expectedRightHash]
+		if leftExist && rightExist {
+			return
+		} else if leftExist {
+			traverseNodes(nodeLeft, path, nodeHashes)
+		} else if rightExist {
+			traverseNodes(nodeRight, path, nodeHashes)
+		}
+	} else if currNode.Edge != nil {
+		edgeNode, exist := nodeHashes[*currNode.Edge.Child]
+		if !exist {
+			return
+		}
+		traverseNodes(edgeNode, path, nodeHashes)
+	}
+}
+
+// Remove duplicates and merges proof paths into a single path
```

Comment:
```suggestion
// MergeProofPaths removes duplicates and merges proof paths into a single path
```
also add "this produces the nodes in specific order - [ common, leftPath, rightPath ] "...
but order is not important hence the `Split...` can discover ... yeah something like this - as we both forget about this detail :P sooner than later